io/src/main/scala/fs2/io/io.scala

```diff
@@ -139,44 +129,26 @@ package object io {
   )(
       f: OutputStream => F[Unit]
   ): Stream[F, Byte] = {
-
-    val mkOutput: Resource[F, (OutputStream, InputStream)] =
-      Resource.make(Sync[F].delay {
-        val os = new PipedOutputStream()
-        val is = new PipedInputStream(os)
-        (os: OutputStream, is: InputStream)
-      })(
-        ois =>
-          blocker.delay {
-            // Piped(I/O)Stream implementations cant't throw on close, no need to nest the handling here.
-            ois._2.close()
-            ois._1.close()
-          }
-      )
-
-    Stream.resource(mkOutput).flatMap {
-      case (os, is) =>
-        Stream.eval(Deferred[F, Option[Throwable]]).flatMap { err =>
-          // We need to close the output stream regardless of how `f` finishes
-          // to ensure an outstanding blocking read on the input stream completes.
-          // In such a case, there's a race between completion of the read
-          // stream and finalization of the write stream, so we capture the error
-          // that occurs when writing and rethrow it.
-          val write = f(os).guaranteeCase(
-            ec =>
-              blocker.delay(os.close()) *> err.complete(ec match {
-                case ExitCase.Error(t) => Some(t)
-                case _                 => None
-              })
-          )
-          val read = readInputStream(is.pure[F], chunkSize, blocker, closeAfterUse = false)
-          read.concurrently(Stream.eval(write)) ++ Stream.eval(err.get).flatMap {
-            case None    => Stream.empty
-            case Some(t) => Stream.raiseError[F](t)
-          }
-        }
+    val mkIn = Resource.fromAutoCloseable(Sync[F].delay { new PipedInputStream })
+    def mkOut(is: PipedInputStream) =
+      Resource.fromAutoCloseable(Sync[F].delay { new PipedOutputStream(is) })
+
+    Stream.resource(mkIn).flatMap { is =>
+      Stream.eval(Deferred[F, Unit]).flatMap { connected =>
+        val write = Stream.eval_(mkOut(is).evalTap(_ => connected.complete(())).use(f))
```

Comment:
I think we want to `blocker.blockOn(write)`. While it's true that the caller *can* do this, I feel like the implementation should, because it's required 100% of the time.